src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -2267,8 +2265,10 @@ PlanRouterQuery(Query *originalQuery,
 	/* we need anchor shard id for select queries with router planner */
 	uint64 shardId = GetAnchorShardId(*prunedShardIntervalListList);
 
-	bool hasLocalRelation = relationRestrictionContext->hasLocalRelation;
-
+	/* both Postgres tables and materialized tables are locally avaliable */
```

Comment:
my current thinking is to  consider :
```C
bool hasLocalRelation =
		rteProperties->hasPostgresLocalTable || rteProperties->hasMaterializedView;
```

What  we are looking for is wether the data is locally avaliable or not on a PG table. Regular views are different because they are already expanded in  the  queryTree,  so any local tables inside the  view will be detected  anyway. For materialized  views, that's not the case.

Does that make sense  @marcocitus ?

Reply:
I think materialized view is oddly selective. What about RELKIND_FOREIGN_TABLE or RELKIND_PARTITIONED_TABLE?

Reply:
I guess those do fall through to hasPostgresLocalTable. 

Is there a way to treat materialized views in the same way? or are they too special?

It's going to be hard to remember these subtle distinctions.

Reply:
Name should probably be hasPostgresLocalRelation

Reply:
> Is there a way to treat materialized views in the same way? or are they too special?

Unfortunately they look special :/ 

For regular tables, partitioned tables or foreign tables, they can be a local or distributed tables and we can qualify them accurately.

For regular views, we don't care because their definitions are already be in the same query tree.

For Mat views, they are just local tables in the queries. But, when `REFRESH 
 MATERIALIZED VIEW` is used, it behaves similar to regular views, adds the definition.

So, still looking for a middle ground

Reply:
got it, thanks for the clarification. Can you add that clarification to a comment somewhere?

Reply:
This is not the best approach to  detect  joins between local tables/mat.view, but seems ok to continue for now.

 We should probably explicitly check the joinTree as opposed to any rangeTable entry in the queryTree. That'd improve the situation a bit. I'll follow up such a change in #4128 as that's already explicitly checking joins

Reply:
For  this  specific point, when user does a `REFRESH MAT VIEW` where the underlying query is a router query, we'll have `rteProperties->hasMaterializedView = true` and router planner would fail. Looking into that:
```SQL
create materialized view router_view as select * from users_table WHERE user_id = 1 FOR UPDATE;
SELECT 7
Time: 133.988 ms
ERROR:  could not run distributed query with FOR UPDATE/SHARE commands
HINT:  Consider using an equality filter on the distributed table's partition column.
Time: 3.502 ms
```

Reply:
Though it is not a regression from 9.4. It is the same at least from Citus  8.3 (the error messages were different: `ERROR:  relation router_view is not distributed`

So, I'm leaning towards leaving this as is and open an issue. What do you think?

Reply:
We should be doing something like: https://github.com/citusdata/citus/commit/48025072431e3946ff342ff3e230840de6e08bbc though I  don't want to change the behavior right before the release.

And, we should probably doing  similar changes in some other places as well

Reply:
(https://github.com/citusdata/citus/issues/4297)